File: super_grasp/src/super_grasp/superquadric.cpp
#include "super_grasp/superquadric.h"
#include <ros/ros.h>
#include <chrono>
#include <random>
#include <cmath>

namespace super_grasp {

Superquadric::Superquadric(std::shared_ptr<open3d::geometry::PointCloud> pcd,
                           int downsample,
                           bool debug)
    : downsample_(downsample), debug_(debug) {

    auto init_time = std::chrono::high_resolution_clock::now();

    // Estimate shape parameters
    auto [e1, e2] = estimateE(pcd);
    e1_ = e1;
    e2_ = e2;

    // Create superquadric mesh and pose
    auto [sq, pose] = createSuperquadric(pcd, e1_, e2_);
    superquadric_ = sq;
    pose_ = pose;

    if (debug_) {
        auto end_time = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - init_time);
        ROS_INFO("Superquadric: time: %.3fs", duration.count() / 1000.0);
        ROS_INFO("Superquadric: Num Points: %zu", superquadric_->points_.size());
    }
}

Eigen::Vector3d Superquadric::calculateKurtosis(const Eigen::MatrixXd& points) {
    // Calculate Fisher kurtosis for each axis
    // Kurtosis = E[(X-μ)^4] / σ^4 - 3 (Fisher/excess kurtosis)
    
    Eigen::Vector3d kurtosis;
    int n = points.rows();
    
    for (int axis = 0; axis < 3; ++axis) {
        Eigen::VectorXd col = points.col(axis);
        
        double mean = col.mean();
        double variance = (col.array() - mean).square().sum() / n;
        double std_dev = std::sqrt(variance);
        
        if (std_dev < 1e-10) {
            kurtosis(axis) = 0.0;
            continue;
        }
        
        // Fourth moment
        double fourth_moment = (col.array() - mean).pow(4).sum() / n;
        
        // Fisher kurtosis (excess kurtosis)
        kurtosis(axis) = (fourth_moment / (variance * variance)) - 3.0;
    }
    
    return kurtosis;
}

std::pair<double, double> Superquadric::estimateE(
    std::shared_ptr<open3d::geometry::PointCloud> pcd) {
    
    try {
        Eigen::MatrixXd points(pcd->points_.size(), 3);
        for (size_t i = 0; i < pcd->points_.size(); ++i) {
            points.row(i) = pcd->points_[i];
        }

        Eigen::Vector3d centroid = pcd->GetCenter();
        Eigen::MatrixXd centered_points = points.rowwise() - centroid.transpose();

        // Compute kurtosis
        Eigen::Vector3d krt = calculateKurtosis(centered_points);

        // Shape parameter along z-axis
        double e1 = std::clamp(1.0 + (krt(2) - 3.0) * 0.1, 0.3, 2.0);

        // Shape parameter along xy-plane
        double e2 = std::clamp(1.0 + ((krt(0) + krt(1)) / 2.0 - 3.0) * 0.1, 0.3, 2.0);

        if (debug_) {
            ROS_INFO("Estimated e1: %.3f, e2: %.3f", e1, e2);
        }

        return {e1, e2};

    } catch (const std::exception& e) {
        ROS_ERROR("superquadric [estimateE] Error: %s", e.what());
        return {1.0, 1.0};
    }
}

std::pair<std::shared_ptr<open3d::geometry::PointCloud>, Eigen::Matrix3d>
Superquadric::createSuperquadric(std::shared_ptr<open3d::geometry::PointCloud> pcd,
                                 double e1, double e2) {
    try {
        // Get point cloud data
        Eigen::MatrixXd points(pcd->points_.size(), 3);
        for (size_t i = 0; i < pcd->points_.size(); ++i) {
            points.row(i) = pcd->points_[i];
        }

        // Check if point cloud is too flat
        Eigen::Vector3d min_pt = points.colwise().minCoeff();
        Eigen::Vector3d max_pt = points.colwise().maxCoeff();
        Eigen::Vector3d ranges = max_pt - min_pt;

        if (debug_) {
            ROS_INFO("Point cloud ranges: X=%.4f, Y=%.4f, Z=%.4f", 
                     ranges(0), ranges(1), ranges(2));
        }

        // If any dimension is too small, add artificial thickness
        double min_range = 0.01; // 1cm
        for (int dim = 0; dim < 3; ++dim) {
            if (ranges(dim) < min_range) {
                if (debug_) {
                    ROS_INFO("Segment is flat in dimension %d, adding artificial thickness", dim);
                }
                
                // Add small noise in the flat dimension
                std::random_device rd;
                std::mt19937 gen(rd());
                std::normal_distribution<> dist(0.0, min_range / 4.0);
                
                for (size_t i = 0; i < pcd->points_.size(); ++i) {
                    pcd->points_[i](dim) += dist(gen);
                }
            }
        }

        // Compute oriented bounding box
        auto obb = pcd->GetOrientedBoundingBox();
        double a1 = obb.extent_(0) / 2.0;
        double a2 = obb.extent_(1) / 2.0;
        double a3 = obb.extent_(2) / 2.0;
        Eigen::Matrix3d R_matrix = obb.R_;
        Eigen::Vector3d center = obb.center_;

        // Decide target number of points
        int n_points = std::max(100, static_cast<int>(pcd->points_.size() * downsample_ / 100));

        // Random number generation
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_real_distribution<> uniform_dist(-1.0, 1.0);

        // Generate superquadric parameters
        std::vector<double> eta(n_points);
        std::vector<double> omega(n_points);
        
        for (int i = 0; i < n_points; ++i) {
            double u = uniform_dist(gen);
            eta[i] = std::asin(std::clamp(u, -1.0, 1.0));
            omega[i] = uniform_dist(gen) * M_PI;
        }

        // Generate superquadric points
        Eigen::MatrixXd V_local(n_points, 3);
        
        for (int i = 0; i < n_points; ++i) {
            double ce = std::cos(eta[i]);
            double se = std::sin(eta[i]);
            double co = std::cos(omega[i]);
            double so = std::sin(omega[i]);

            double x = a1 * sgn(ce) * std::pow(std::abs(ce), e1) * 
                       sgn(co) * std::pow(std::abs(co), e2);
            double y = a2 * sgn(ce) * std::pow(std::abs(ce), e1) * 
                       sgn(so) * std::pow(std::abs(so), e2);
            double z = a3 * sgn(se) * std::pow(std::abs(se), e1);

            V_local.row(i) << x, y, z;
        }

        // Transform to world coordinates
        Eigen::MatrixXd V_world = (R_matrix * V_local.transpose()).transpose();
        V_world.rowwise() += center.transpose();

        // Build point cloud
        auto sq_pcd = std::make_shared<open3d::geometry::PointCloud>();
        sq_pcd->points_.resize(n_points);
        
        for (int i = 0; i < n_points; ++i) {
            sq_pcd->points_[i] = V_world.row(i);
        }

        // Estimate normals
        sq_pcd->EstimateNormals(
            open3d::geometry::KDTreeSearchParamHybrid(0.02, 30));

        return {sq_pcd, R_matrix};

    } catch (const std::exception& e) {
        ROS_ERROR("superquadric [createSuperquadric] Error: %s", e.what());
        return {std::make_shared<open3d::geometry::PointCloud>(), Eigen::Matrix3d::Identity()};
    }
}

} // namespace super_grasp

File: super_grasp/src/super_grasp/superquadric.py
#!/usr/bin/env python3.8

"""
    superquadric: estimates and firts superquadrics to partial point cloud model, 
                  shape values e1, e2 are estimated based on kurtosis of point cloud
                  size values a1, a2, a3 are estimated based on bounding box of point cloud
                  pose is estimated based on oriented bounding box of point cloud

    Input:
        pcd: point cloud of the object to fit superquadric to
        debug: if True, print debug information
    
    Output:
        superquadric: point cloud of the fitted superquadric
        pose: 3x3 rotation matrix of the superquadric  

    Developed by: Jayden Kanbour as part of undergraduate thesis for UNSW Computer Science and Engineering
    Date: 26th November 2025
    Email: jkanbour1@gmail.com
    UNSW Student Id: z5316799

"""

import numpy as np
import open3d as o3d
from scipy.stats import kurtosis
import time

class Superquadric:
    def __init__(self, pcd, downsample=30, debug=False):

        init_time = time.time()
        self.debug = debug
        self.downsample = downsample
        self.extent = None
        self.pose = None
        self.center = None

        # estimate e1, e2 for superquadric fitting
        self.e1, self.e2 = self.estimateE(pcd)

        # create superquadric mesh and pose
        self.superquadric = self.createSuperquadric(pcd, self.e1, self.e2)

        if self.debug:
            print(f"Superquadric: time: {time.time() - init_time:.3f}s")
            print(f"Superquadric: SQ time: {time.time() - init_time:.3f}s")
            print(f"Superquadric: Num Points: {len(self.superquadric.points)}")


    def estimateE(self, pcd):

        """
            Estimate superquadric shape parameters e1 and e2 based on point cloud kurtosis.
            
            e1 controls shape along z-axis (elongation/flattening)
            e2 controls shape in xy-plane (roundness/squareness)
        """

        try:
            points = np.asarray(pcd.points)
            center = pcd.get_center()
            centered_points = points - center

            # Compute Fisher kurtosis for x, y, z axes of the point cloud
            krt = kurtosis(centered_points, axis=0, fisher=True, bias=False)

            # Shape parameter along z-axis based on kurtosis (controls superquadric elongation or flattening)
            e1 = np.clip(1 + (krt[2] - 3) * 0.1, 0.3, 2.0)

            # Shape parameter along xy-plane based on average x and y kurtosis
            e2 = np.clip(1 + ((krt[0] + krt[1]) / 2 - 3) * 0.1, 0.3, 2.0)

            if self.debug:
                print(f"Estimated e1: {e1:.3f}, e2: {e2:.3f}")

            return e1, e2
        
        except Exception as e:
            print(f"superquadric [estimateE] Error: {e}")
            return 1.0, 1.0

    
    def createSuperquadric(self, pcd, e1, e2):
        """
            Create a superquadric point cloud fitted to the input point cloud.
            Uses oriented bounding box of the point cloud to determine size and pose.
        """

        try:
            # Check if point cloud is too flat
            points = np.asarray(pcd.points)
            ranges = points.max(axis=0) - points.min(axis=0)
            
            if self.debug:
                print(f"Point cloud ranges: X={ranges[0]:.4f}, Y={ranges[1]:.4f}, Z={ranges[2]:.4f}")
            
            # If any dimension is too small, add artificial thickness
            min_range = 0.01  # 1cm
            if np.any(ranges < min_range):
                flat_dim = np.argmin(ranges)
                if self.debug:
                    print(f"Segment is flat in dimension {flat_dim}, adding artificial thickness")
                
                # Add small noise in the flat dimension
                noise = np.zeros_like(points)
                noise[:, flat_dim] = np.random.normal(0, min_range/4, len(points))
                pcd.points = o3d.utility.Vector3dVector(points + noise)
            
            # Now compute OBB
            obb = pcd.get_oriented_bounding_box()

            self.extent = obb.extent
            a1, a2, a3 = self.extent[0] / 2.0, self.extent[1] / 2.0, self.extent[2] / 2.0

            self.pose = obb.R

            self.center = obb.center

            # Decide target number of points for final model (downsampling if needed)
            n_points = max(100, int(len(pcd.points) * self.downsample // 100))

            # Superquadric parameter generation x, y, z
            eta = (np.random.rand(n_points) - 0.5) * np.pi 
            eta *= 1.0
            eta = eta / 1.0 
            u = np.random.rand(n_points) * 2.0 - 1.0 
            eta = np.arcsin(np.clip(u, -1.0, 1.0))  
            omega = (np.random.rand(n_points) * 2.0 - 1.0) * np.pi 

            def sgn(x):
                return np.sign(x + 1e-15)

            ce, se = np.cos(eta), np.sin(eta)
            co, so = np.cos(omega), np.sin(omega)

            x = a1 * sgn(ce) * np.abs(ce)**e1 * sgn(co) * np.abs(co)**e2
            y = a2 * sgn(ce) * np.abs(ce)**e1 * sgn(so) * np.abs(so)**e2
            z = a3 * sgn(se) * np.abs(se)**e1

            # World transform
            V_local = np.stack([x, y, z], axis=1) 
            V_world = (self.pose @ V_local.T).T + self.center

            # Build point cloud directly
            sq_pcd = o3d.geometry.PointCloud()
            sq_pcd.points = o3d.utility.Vector3dVector(V_world.astype(np.float64))

            # Estimate normals once
            sq_pcd.estimate_normals(
                search_param=o3d.geometry.KDTreeSearchParamHybrid(radius=0.02, max_nn=30)
            )

            return sq_pcd
        
        except Exception as e:
            print(f"superquadric [createSuperquadric] Error: {e}")
            return 1.0, 1.0

    def getSuperquadricMesh(self):
        return self.superquadric
    
    def getCenter(self):
        return self.center
    
    def getBBOXExtent(self):
        return self.extent
    
    def getSQPose(self):
        return self.pose